File: mo_lucro_app/lib/services/goal_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/goal_model.dart';

class GoalService {
  final _db = Supabase.instance.client;

  String get _userId {
    final user = _db.auth.currentUser;
    if (user == null) throw Exception('Usuário não autenticado');
    return user.id;
  }

  Future<List<GoalModel>> getGoals() async {
    final res = await _db
        .from('goals')
        .select()
        .eq('user_id', _userId)
        .order('created_at', ascending: false);

    return (res as List)
        .map((row) => GoalModel.fromMap(row as Map<String, dynamic>))
        .toList();
  }

  Future<void> addGoal({
    required String title,
    required double targetValue,
    double currentValue = 0,
    DateTime? deadline,
  }) async {
    final payload = <String, dynamic>{
      'user_id': _userId,
      'name': title,
      'target': targetValue,
      'current': currentValue,
    };
    if (deadline != null) {
      payload['deadline'] = deadline.toIso8601String();
    }
    await _db.from('goals').insert(payload);
  }

  Future<void> updateGoalProgress(String id, double currentValue) async {
    await _db
        .from('goals')
        .update({'current': currentValue}).eq('id', id);
  }

  Future<void> deleteGoal(String id) async {
    await _db.from('goals').delete().eq('id', id);
  }
}

File: mobile/src/services/goalService.ts
import { goalFromRow, type Goal } from '@/models';
import { requireUserId, supabase } from '@/lib/supabase';

/** Porte de lib/services/goal_service.dart. */

export async function getGoals(): Promise<Goal[]> {
  const userId = await requireUserId();

  const { data, error } = await supabase
    .from('goals')
    .select()
    .eq('user_id', userId)
    .order('created_at', { ascending: false });

  if (error) throw error;
  return (data ?? []).map(goalFromRow);
}

export async function addGoal(params: {
  title: string;
  targetValue: number;
  currentValue?: number;
  deadline?: Date | null;
}): Promise<void> {
  const userId = await requireUserId();

  const payload: Record<string, any> = {
    user_id: userId,
    name: params.title,
    target: params.targetValue,
    current: params.currentValue ?? 0,
  };
  if (params.deadline) {
    payload.deadline = params.deadline.toISOString();
  }

  const { error } = await supabase.from('goals').insert(payload);
  if (error) throw error;
}

export async function updateGoalProgress(id: string, currentValue: number): Promise<void> {
  const { error } = await supabase.from('goals').update({ current: currentValue }).eq('id', id);
  if (error) throw error;
}

export async function deleteGoal(id: string): Promise<void> {
  const { error } = await supabase.from('goals').delete().eq('id', id);
  if (error) throw error;
}
